### DISCUSSION TITLE
Workflow with stacked branches

### DISCUSSION THREAD

**Discussion by GiorgosAlexakis** [arxanas/git-branchless#745] [Q&A]
Consider four branches. The second one is stacked on top of the first one, the third one is stacked on top of the second and so forth. Each branch can have multiple commits.

Now let's say I checkout to the first commit of the first branch. I add the changes I want but instead of amending the commit I want to create a new commit that will be placed after the first commit of the first branch.

If a new commit is created while I am checked out at the first commit, the rest of the branches won't include this added commit. One obvious solution that doesn't involve git-branchless requires you to rebase each branch, i.e go to second branch git rebase first_branch and so forth.

How can I use git-branchless to avoid doing all this manual work?

**Comment by arxanas** [comment]
In the latest development version, instead of `git commit`, you can use `git record --insert` (not documented in the Wiki), or run `git move -s 'siblings(.)'` immediately after making the commit.

In the latest released version, the `siblings` function doesn't yet exist, so you'll have to write `git move -s <hash>` or `git move -s 'children(parents(.)) - @'` instead.

**Reply by GiorgosAlexakis** [reply]
Thanks I will try it! If I then want to update the remote branches can I do?:

```bash
# $reference is the name of the first branch of the stack
git submit 'stack($reference)'
# I usually confirm the list of branches I want to update by looking at the result of this:
git query -b "branches() & stack($reference)"
```

For anyone interested in automatically publishing the branches and creating PRs for them in Github:

```bash
# enable this 
git config --global --add --bool push.autoSetupRemote true

# confirm the list of branches you want to publish
git query -b "branches() & stack($reference)"

# use git submit with --create flag
git submit 'stack($reference)' --create

# create stack alias
git config --global alias.crst '!f() { git submit "stack($1)" --create; }; f'
# usage
git crst $reference

# update stack alias
git config --global alias.updst '!f() { git submit "stack($1)"; }; f'
# usage
git updst $reference 

# make sure gh cli is installed
# review stack alias `rvst` that automatically creates a PR for each published branch in your stack
git config --global alias.rvst '!f() { branches=$(git query -b "branches() & stack($1)"); if [ -z "$branches" ]; then echo "No branch in stack"; 
else for branch in $branches;
do gh pr create --base main --fill --assignee @me --head $branch; done; fi; }; f'
# usage
git rvst $reference
# you can obviously customize the above alias to prompt terminal to ask you the name of the PR for each branch etc..
# I tend to create a PR as I create the branches though so I haven't really used the last one.
```

**Reply by arxanas** [reply]
> Thanks I will try it! If I then want to update the remote branches can I do?:

Yep, that should work.

**Reply by GiorgosAlexakis** [reply]
What happens if I use `git record --insert` and conflicts occur?

**Reply by arxanas** [reply]
It should succeed in committing, but abort the rebase and tell you to run `git move --merge - s 'siblings(.)' `.
